mira/dkg/client.py

```diff
@@ -217,6 +217,24 @@ def get_node_counter(self) -> Counter:
             {label: self.query_tx(f"MATCH (n:{label}) RETURN count(*)")[0][0] for label in labels}
         )
 
+    def search(self, query: str, limit: int = 25) -> List[Entity]:
+        """Search nodes for a given name substring."""
+        query_lower = query.lower().replace("-", "").replace("_", "")
+        cypher = dedent(
+            f"""\
+            MATCH (n) 
+            WHERE 
+                replace(replace(toLower(n.name), '-', ''), '_', '') CONTAINS '{query_lower}'
+                OR any(
+                    synonym IN n.synonyms 
+                    WHERE replace(replace(toLower(synonym), '-', ''), '_', '') CONTAINS '{query_lower}'
+                )
+            RETURN n 
+            LIMIT {limit}
+        """
+        )
+        return [Entity(**n) for n in self.query_nodes(cypher)]
```

Comment:
I am wondering if we should try to do some sorting/ranking at this point - perhaps a simple string similarity score to rank results?

Reply:
That's possible using the built-in diff library or something external like `fuzzywuzzy`. Looking forward, there are for sure a lot of things to be done to improve the goodness of these (some task queues like Redis have support for this, elastic search, etc.)

Reply:
Yes, I think we could use fuzzywuzzy, I think it's a simple and useful first approach to get a scored ranking.

Reply:
Done in 6db4a52 with built-in `difflib`